include/rbs/util/rbs_constant_pool.h

```diff
@@ -0,0 +1,94 @@
+/**
+ * @file rbs_constant_pool.h
+ *
+ * A data structure that stores a set of strings.
+ *
+ * Each string is assigned a unique id, which can be used to compare strings for
+ * equality. This comparison ends up being much faster than strcmp, since it
+ * only requires a single integer comparison.
+ */
+#ifndef RBS_CONSTANT_POOL_H
+#define RBS_CONSTANT_POOL_H
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+/**
+ * When we allocate constants into the pool, we reserve 0 to mean that the slot
+ * is not yet filled. This constant is reused in other places to indicate the
+ * lack of a constant id.
+ */
+#define RBS_CONSTANT_ID_UNSET 0
+
+/**
+ * A constant id is a unique identifier for a constant in the constant pool.
+ */
+typedef uintptr_t rbs_constant_id_t;
+
+/** A constant in the pool which effectively stores a string. */
+typedef uintptr_t rbs_constant_t;
+
+/** The overall constant pool, which stores constants found while parsing. */
+typedef struct {
+    void *dummy; // Workaround for structs not being allowed to be empty.
```

Comment:
Eventually, this will be replaced with the real storage that will store the constant pool, instead of the one built into the Ruby VM.